sycl/plugins/level_zero/usm_allocator.cpp

```diff
@@ -414,24 +414,35 @@ class Bucket {
   // routines, slab map and etc.
   USMAllocContext::USMAllocImpl &OwnAllocCtx;
 
+  // For counting number of slabs in the pool.
```

Comment:
What does the "pool" here refers to? (I only see Buckets, Slabs in the Buckets, and Chunks in the Slabs).
For what purpose do you need to count "chunked" Slabs? Why total # of Slabs is not enough?
Please add a high-level design notes in the description of this PR.

Reply:
An L0 allocation is called a "slab". The pool consists of L0 allocations (slabs) that are not in use by the program.

Allocations are in different "buckets", based on their size. Each bucket manages a set of slabs in a size range.

For larger user allocations, an entire slab is used for a single user allocation. For these allocation sizes, slabs move between two lists: Unavailable (in use by program) and Available (in the pool). The size of the Available list indicates how many slabs in that bucket are "in the pool".

Smaller allocations are "chunked" out of individual slabs. More than one user allocation is in a slab. Here, an Unavailable slab is one all of whose chunks are in use by the program, and an Available slab is one with some chunks not in use by the program and available for allocation. Thus, the slabs in the Available list are not in the pool according to the definition above, as these slabs have some chunks in use by the program and some chunks available for allocation. To implement pooling of chunked slabs we allow one slab in the Available list to be completely empty, in effect allowing a pool size of 1 for chunked allocations.

For chunked buckets, the counter chunkedSlabsInPool counts how many of the Available slabs in a bucket are completely unused, i.e. in the pool by our definition.

Reply:
Thanks for the short summary of the representation!
Please add it as a source code comment (maybe at a class level).

> Available (in the pool)
...
the slabs in the Available list are not in the pool according to the definition above

I find these being in conflict, which one is true?

> To implement pooling of chunked slabs we allow one slab in the Available list to be completely empty, in effect allowing a pool size of 1 for chunked allocations. For chunked buckets, the counter chunkedSlabsInPool counts how many of the Available slabs in a bucket are completely unused, i.e. in the pool by our definition.

I still don't see why you need to do extra counting to re-use small/chunked allocations.

Reply:
For allocations that use an entire slab each, the entries in the Available list are entries in the pool. Each slab is available for a new allocation. The size of the Available list is the size of the pool.

For allocations that use slabs in chunked mode, slabs will be in the the Available list if any _one or more_ of their chunks is free. The entire slab is not necessarily free, just some chunks in the slab are free. To implement pooling we will allow _one_ slab in the Available list to be entirely empty. Normally such a slab would have been freed from USM. But now we don't, and treat this slab as "in the pool".

When a slab becomes entirely free we have to decide whether to return it to USM or keep it allocated. A simple check for size of the Available list is not sufficient to check whether any slab has been pooled yet. We would have to traverse the entire Available list and check if any of them is entirely free. Instead we keep a counter of entirely empty slabs within the Available list. to speed up the process of checking if a slab in this bucket is already pooled.